extensions/python/PythonDependencyInstaller.cpp

```diff
@@ -0,0 +1,167 @@
+    logger_->log_debug("Python processor dir was specified at: {}", config_state_.python_processor_dir.string());
+  } else {
+    logger_->log_debug("No valid python processor dir was not specified in properties");
+  }
+  createVirtualEnvIfSpecified();
+  addVirtualenvToPath();
+}
+
+std::vector<std::filesystem::path> PythonDependencyInstaller::getRequirementsFilePaths() const {
+  if (!std::filesystem::exists(config_state_.python_processor_dir)) {
+    return {};
+  }
+  std::vector<std::filesystem::path> paths;
+  for (const auto& entry : std::filesystem::recursive_directory_iterator(std::filesystem::path{config_state_.python_processor_dir})) {
+    if (std::filesystem::is_regular_file(entry.path()) && entry.path().filename() == "requirements.txt") {
+      paths.push_back(entry.path());
+    }
+  }
+  return paths;
+}
+
+void PythonDependencyInstaller::createVirtualEnvIfSpecified() const {
+  if (config_state_.virtualenv_path.empty()) {
+    return;
+  }
+  if (!std::filesystem::exists(config_state_.virtualenv_path) || std::filesystem::is_empty(config_state_.virtualenv_path)) {
+    logger_->log_info("Creating python virtual env at: {}", config_state_.virtualenv_path.string());
+    auto venv_command = "\"" + config_state_.python_binary + "\" -m venv \"" + config_state_.virtualenv_path.string() + "\"";
+    auto return_value = std::system(encapsulateCommandInQuotesIfNeeded(venv_command).c_str());
+    if (return_value != 0) {
+      throw PythonScriptException(fmt::format("The following command creating python virtual env failed: '{}'", venv_command));
+    }
+  }
+}
+
+void PythonDependencyInstaller::installDependenciesFromRequirementsFiles() const {
+  std::string automatic_install_str;
+  if (!config_state_.isPackageInstallationNeeded()) {
+    return;
+  }
+  auto requirement_file_paths = getRequirementsFilePaths();
+  for (const auto& requirements_file_path : requirement_file_paths) {
+    logger_->log_info("Installing python packages from the following requirements.txt file: {}", requirements_file_path.string());
+    std::string pip_command;
+#if WIN32
+    pip_command.append("\"").append((config_state_.virtualenv_path / "Scripts" / "activate.bat").string()).append("\" && ");
+#else
+    pip_command.append(". \"").append((config_state_.virtualenv_path / "bin" / "activate").string()).append("\" && ");
+#endif
+    pip_command.append("\"").append(config_state_.python_binary).append("\" -m pip install --no-cache-dir -r \"").append(requirements_file_path.string()).append("\"");
+    auto return_value = std::system(encapsulateCommandInQuotesIfNeeded(pip_command).c_str());
+    if (return_value != 0) {
+      throw PythonScriptException(fmt::format("The following command to install python packages failed: '{}'", pip_command));
+    }
+  }
+}
+
+void PythonDependencyInstaller::evalScript(std::string_view script) {
+  GlobalInterpreterLock gil;
+  const auto script_file = "# -*- coding: utf-8 -*-\n" + std::string(script);
```

Comment:
We could save a large string copy by using `utils::string::join_pack`. Even though the temporary string is an rvalue, the concatenated result can't reuse its buffer.

Reply:
Good point, updated in 99359a0eda994fd5efa3095ebb76811ff1136f05